lib/AST/UnqualifiedLookup.cpp

```diff
@@ -0,0 +1,1806 @@
+  case ASTScopeKind::IfStmt:
+  case ASTScopeKind::GuardStmt:
+  case ASTScopeKind::RepeatWhileStmt:
+  case ASTScopeKind::ForEachStmt:
+  case ASTScopeKind::DoCatchStmt:
+  case ASTScopeKind::SwitchStmt:
+  case ASTScopeKind::Accessors:
+  case ASTScopeKind::TopLevelCode:
+    llvm_unreachable("no local declarations?");
+
+  case ASTScopeKind::ExtensionGenericParams:
+  case ASTScopeKind::GenericParams:
+    return DeclVisibilityKind::GenericParameter;
+
+  case ASTScopeKind::AbstractFunctionParams:
+  case ASTScopeKind::Closure:
+  case ASTScopeKind::PatternInitializer: // lazy var 'self'
+    return DeclVisibilityKind::FunctionParameter;
+
+  case ASTScopeKind::AfterPatternBinding:
+  case ASTScopeKind::ConditionalClause:
+  case ASTScopeKind::ForEachPattern:
+  case ASTScopeKind::BraceStmt:
+  case ASTScopeKind::CatchStmt:
+  case ASTScopeKind::CaseStmt:
+    return DeclVisibilityKind::LocalVariable;
+  }
+
+  llvm_unreachable("Unhandled ASTScopeKind in switch.");
+}
+
+
+
+namespace {
+
+/// Determine whether unqualified lookup should look at the members of the
+/// given nominal type or extension, vs. only looking at type parameters.
+template <typename D> bool shouldLookupMembers(D *decl, SourceLoc loc) {
+  // Only look at members of this type (or its inherited types) when
+  // inside the body or a protocol's top-level 'where' clause. (Why the
+  // 'where' clause? Because that's where you put constraints on
+  // inherited associated types.)
+
+  // When we have no source-location information, we have to perform member
+  // lookup.
+  if (loc.isInvalid() || decl->getBraces().isInvalid())
+    return true;
+
+  // Within the braces, always look for members.
+  auto &ctx = decl->getASTContext();
+  if (ctx.SourceMgr.rangeContainsTokenLoc(decl->getBraces(), loc))
+    return true;
+
+  // Within 'where' clause, we can also look for members.
+  if (auto *whereClause = decl->getTrailingWhereClause()) {
+    SourceRange whereClauseRange = whereClause->getSourceRange();
+    if (whereClauseRange.isValid() &&
+        ctx.SourceMgr.rangeContainsTokenLoc(whereClauseRange, loc)) {
+      return true;
+    }
+  }
+
+  // Don't look at the members.
+  return false;
+}
+} // end anonymous namespace
+
+namespace {
+class LegacyUnqualifiedLookup;
+
+class UnqualifiedLookupFactory {
+public:
+  using Flags = UnqualifiedLookup::Flags;
+  using Options = UnqualifiedLookup::Options;
+  
+private:
+
+  // TODO: better name than DC
+  struct DCAndResolvedIsCascadingUse {
+    DeclContext *const DC;
+    const bool isCascadingUse;
+  };
+
+  /// Finds lookup results based on the types that self conforms to.
+  /// For instance, self always conforms to a struct, enum or class.
+  /// But in addition, self could conform to any number of protocols.
+  /// For example, when there's a protocol extension, e.g. extension P where
+  /// self: P2, self also conforms to P2 so P2 must be searched.
+  class ResultFinderForSelfsConformances {
```

Comment:
Rather than "for Self's conformances", maybe we want to say this is for "for type contexts"? That's the general case; that we're collecting constraints on `Self` to find these types is more of an implementation detail.

Reply:
Good thought. As it is, I believe that the structure is self-specific because the constructer calls `findNominalTypesThatSelfMustConformTo` which in turn calls `dc->getSelfNominalTypeDecl()`. But yes, it could certainly be parameterized and generalized, by for example passing in a `NominalTypeDecl` in to the constructor. Would there be a use for such a generalized facility?